src/controllers/schedule.js

```diff
@@ -0,0 +1,281 @@
+                        let intEnd = lastEndTime.clone();
+                        intEnd.hour(endHr).minute(0).second(0).millisecond(0);
+                        freeInts.push([lastEndTime.format(), intEnd.format()]);
+                    }
+                    // Check if the end is within hours of operation
+                    if (validTime(busyInt[0]) == 0) {
+                        console.log("\tends within hours of operation")
+                        let intStart = busyInt[0].clone();
+                        intStart.hour(startHr).minute(0).second(0).millisecond(0);
+                        freeInts.push([intStart.format(), busyInt[0].format()]);
+                    }
+                    // If neither are within hours of operation, do nothing!
+                }
+                // If the interval ends greater than a day in the future...
+                else if (busyInt[0].date() - lastEndTime.date() > 1) {
+                    // Then we can add the whole day
+                    let intStart = lastEndTime.clone()
+                    intStart.hour(startHr).minute(0).second(0).millisecond(0);
+                    let intEnd = intStart.clone();
+                    intEnd.hour(endHr).minute(0).second(0).millisecond(0);
+
+                    freeInts.push([intStart.format(), intEnd.format()]);
+                }
+            }
+            // We can't skip any days!
+            if (busyInt[0].date() > lastEndTime.date() + 1) {
+                lastEndTime.add(1, "days");
+                lastEndTime.hour(startHr).minute(0).second(0).millisecond(0);
+                console.log('rewind a day!');
+            }
+            else
+                lastEndTime = busyInt[1];
+        });
+        // TODO: I don't know if this function gets the very last free interval (up to the last due date)
+
+        // Print for debugging purposes
+        console.log("FREE INTERVALS");
+        freeInts.forEach((freeInt) => {
+            console.log(freeInt);
+        });
+        console.log();
+
+        // TODO: Make it so that freeInts is already in seconds since epoch so we don't have to convert
+        freeInts = freeInts.map((interval) => {
+            return [
+                moment(interval[0]).unix(), 
+                moment(interval[1]).unix()
+            ];
+        });
+
+        function isValidTask(task) {
+            const due = moment(task.due_date).unix();
+            const duration = task.duration;
+            const start = task.scheduled_time;
+            return start + duration <= due;
+        }
+
+        //  Arguments:
+        //      tasks - An array of task objects, as defined by the task model. Sorted from earliest
+        //          to latest due date.
+        //      freeTime - An array of two-element arrays, each representing an interval of free time.
+        //          The times should be in seconds since epoch.
+        function solve(tasks, freeTime) {
+            if (tasks.length == 0)
```

Comment:
Additionally, this whole flow (schedule tasks given free times) should be its own function in `ScheduleService.js`